moveit_experimental/moveit_jog_arm/include/moveit_jog_arm/make_shared_from_pool.h

```diff
@@ -36,25 +36,21 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *******************************************************************************/
 
-// Server node for arm jogging with MoveIt.
-
 #pragma once
 
-#include "jog_interface_base.h"
+#include <boost/pool/pool_alloc.hpp>
 
-namespace moveit_jog_arm
+namespace moveit
+{
+namespace util
 {
-/**
- * Class JogROSInterface - Instantiated in main(). Handles ROS subs & pubs and creates the worker threads.
- */
-class JogROSInterface : protected JogInterfaceBase
+// Useful template for creating messages from a message pool
```

Comment:
Interesting feature. Where is the pool (size) defined?
Is such a feature available from C++14's standard lib as well? The pool allocator is the only reason to pull in boost and the ROS community strives to replace boost with std features.

Reply:
The pool is allocated by boost as needed using a singleton per type.  Because I used boost::allocate_shared it only does one allocation (both the message and the control block of the shared pointer).  I used a tool to print when malloc is being called and on my machine, the first call to this allocates space for 32 of that type.  After that it will double that storage if I ever am using all 32 and ask for another, and so on and so forth.  The big win here is that you can have all the goodness of the zero-copy from roscpp without the performance hit from memory fragmentation and system calls in every loop iteration (assuming the subscriber doesn't hold onto a ton of them).

C++14 doesn't offer any equivalent functionality to this as far as I know.

Reply:
Also, from my reading of roscpp it seems that the zero-copy implementation depends on a boost::shared_ptr type, not that it couldn't be implemented using std::shared_ptr but that's what I found.

I generally agree that replacing boost with std features where appropriate is the right call.  For now, this functionality doesn't exist in the standard library.

Reply:
Here is my conversation with @gavanderhoorn about this: https://answers.ros.org/question/353198/pre-allocation-of-messages-for-nodelets/

Here is the documentation on boost::pool_allocator:  https://www.boost.org/doc/libs/1_73_0/libs/pool/doc/html/boost/pool_allocator.html

I used the further specialization `fast_pool_allocator` which is optimized for single allocations (think list and not vector) as I'm never asking for an array of new messages.

Reply:
`boost` is doing some magic with the type passed in here: `using allocator_t = boost::fast_pool_allocator<boost::shared_ptr<T>>;`  That type is used as a tag for a singleton pool of memory.  They can't deduce the size from this type and instead it is set the first time I create a new chunk from it on the next line.

I don't like adding complexity just for the fun of it.  I think in this case the tradeoffs you get from using this for allocating messages you send through ROS on the same process are completely worth it as it removes malloc from your loops most of the time (ideally only on your first iteration), solves memory fragmentation from calling malloc for messages inside loops, and the interface I created here looks very similar to the `make_shared` everyone is used to already.

Reply:
I'm all fine with this and already marked the conversation as resolved.
Just a pity that the C++ standard library doesn't support this yet.

Reply:
Unrelated, what are you thoughts on switching to c++17 in master only (and maybe Noetic as it would be a real stretch to install Noetic on 16.04)?

Reply:
Noetic still sticks with C++14:
https://www.ros.org/reps/rep-0003.html#noetic-ninjemys-may-2020-may-2025

Reply:
That's sad... all modern compilers (even MVCPP) fully supports c++17 now.